compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -1029,23 +1102,52 @@ impl<'a> PartialEvaluator<'a> {
         }
     }
 
-    fn try_eval_block(&mut self, block_id: BlockId) -> Result<Value, ()> {
-        self.visit_block(block_id);
-        if self.errors.is_empty() {
-            Ok(self.eval_context.get_current_scope().last_expr_value())
+    fn try_eval_block(&mut self, block_id: BlockId) -> Result<ValueControlFlow, Error> {
+        let block = self.get_block(block_id);
+        let mut return_stmt_id = None;
+        let mut last_control_flow = ValueControlFlow::new_continue(Value::unit());
+
+        // Iterate through the statements until we hit a return or reach the last statement.
+        let mut stmts_iter = block.stmts.iter();
+        for stmt_id in stmts_iter.by_ref() {
+            last_control_flow = self.try_eval_stmt(*stmt_id)?;
+            if matches!(last_control_flow.kind, ControlFlowKind::Return) {
+                return_stmt_id = Some(*stmt_id);
+                break;
+            }
+        }
+
+        // While we support multiple returns within a callable, disallow situations in which non qubit release
+        // statements are left unprocessed.
+        let non_qubit_relase_stmts_remain =
+            stmts_iter.any(|stmt_id| !self.is_qubit_release_stmt(*stmt_id));
+        if non_qubit_relase_stmts_remain {
+            let return_stmt =
+                self.get_stmt(return_stmt_id.expect("a return statement ID must have been set"));
+            Err(Error::Unexpected(
+                "early return".to_string(),
+                return_stmt.span,
+            ))
```

Comment:
Something seems odd to me here... shouldn't it only be early returns in a dynamic context that get treated as error? If the return is due to purely classical control flow, shouldn't it be allowed? For example:
```qsharp
operation XorMore(q : Qubit, cond: Bool) : Unit {
    if cond {
        X(q);
        return;
    }
    H(q);
    if M(q) == One {
        X(q);
    }
}
```
If `XorMore` is called with a classical true, it will perform just `X` and not include any control flow in the RIR, and with a classical false it will have the `H` and one branch in the RIR based on `M`. Both of those could be handled with the infra you have here, but it looks like that would be rejected by this logic.

Reply:
It's possible I'm just missing the intent here, so happy to sync up more on it Monday.

Reply:
This is a good point. This implementation was overly restrictive.

I've changed it such that we only raise an error when we are performing evaluation within a branch (at the callable scope) level. I did this with simple reference counting for now, but it gives us the points where we should handle early returns in a later change.

Happy to sync up on the details :)